feat(iota-framework): Introduce committee selection framework with TopStakers

piotrm50 opened the pull request:
# Description of change

This PR introduces a modified committee selection framework, which allows to easily modify the logic to select the committee members from the list of all active validators. This is duplicate of #3903 which targets `develop` branch directly. 

## Links to any relevant issues

#2763 

## Type of change

Choose a type of change, and delete any options that are not relevant.

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How the change has been tested

Unit tests, running local network. 

## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes

### Release Notes

<!--
Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.
For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates.
-->

- [x] Protocol: Changes the way consensus committee is formed from all active validators who met criteria, to a subset of them with the most stake. 
- [ ] Nodes (Validators and Full nodes):
- [ ] gRPC:
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:

hmoog committed:
migrated first changes

hmoog committed:
fixed errors

hmoog committed:
removed unnecessary code

hmoog committed:
fixed linter remarks

hmoog committed:
started adding reasons to compatibility check

hmoog committed:
removed debug prints

hmoog committed:
increased reporting for compatibility checks

hmoog committed:
cleaned code

hmoog committed:
cleaned code

hmoog committed:
cleaned code

hmoog committed:
cleaned code

hmoog committed:
moved logic to module

hmoog committed:
generated files

hmoog committed:
added select function

piotrm50 committed:
Revert unnecessary changes.

piotrm50 committed:
Generate new baselines for graphql-e2e tests

alexsporn commented:
Please rebase on top of the rebased `core-protocol/procotol-version-4` branch

piotrm50 committed:
Refactor ValidatorSetV2 to avoid copying an entire file.

piotrm50 committed:
Rename active_validator to eligible_validator

piotrm50 committed:
Fix move tests

piotrm50 committed:
Update all snapshots

piotrm50 committed:
Update baselines

piotrm50 committed:
Adjust advance_epoch logic to the new committee selection logic.

piotrm50 committed:
Fix iota-framework-tests

piotrm50 committed:
Regenerate snapshots

piotrm50 committed:
Update postgres test baselines

piotrm50 committed:
Revert modification of public function

piotrm50 committed:
Update postgres test baselines

piotrm50 committed:
Separate emitting events from committee selection.

piotrm50 committed:
Implement TopStakers strategy with tests. Introduce `SystemParametersV2` with `committee_members_count`

piotrm50 committed:
Update baselines

piotrm50 committed:
Fix clippy

piotrm50 committed:
Revert `active` -> `eligible` rename

piotrm50 committed:
Add new values to SystemStateSummary

piotrm50 committed:
remove unnecessary TODO

piotrm50 committed:
Fix tests

Thoralf-M commented on crates/iota-framework/packages/iota-system/tests/validator_set_tests.move:
```diff
@@ -79,17 +96,128 @@ module iota_system::validator_set_tests {
+
     #[test]
     fun test_reference_gas_price_derivation() {
+        // TODO: implement tests that check that gas price is correctly calculated based on committee and not all validators
```
Want to do this in this PR? (Just noticed it's still a DRAFT)

piotrm50 committed:
Move active_validators out of self before consuming it

piotrm50 committed:
Fix clippy

piotrm50 committed:
Update all snapshots

piotrm50 committed:
Update iota-framework-snapshot

piotrm50 committed:
Update populated genesis snapshot for tests

muXxer commented on crates/iota-framework/Cargo.toml:
```diff
@@ -9,6 +9,7 @@ description = "Move framework for iota platform"
 
 [dependencies]
 # external dependencies
+anyhow.workspace = true
```
I don't see where this is used now in the introduced changes.

piotrm50 replied:
True, most likely some leftover from one of PCR2 rebases.

muXxer commented on crates/iota-framework/packages/iota-framework/sources/test/test_utils.move:
```diff
@@ -19,6 +19,92 @@ module iota::test_utils {
+            std::debug::print(&len2);
+            abort(0)
+        };
+
```
Was this move file formatted with the new [move-pretier-plugin](https://marketplace.visualstudio.com/items?itemName=mysten.prettier-move)?

muXxer replied:
Please do that as well for all the other move files your touched :trollface:

Or should we rather do it at the end, after everyone reviewed and approved everything, to keep the changes small for now?

piotrm50 replied:
I'll see how much changes it generates. But maybe it would be easier to do it at the end if there's a lot of changes on unrelated lines.

Ok, it generates changes on the whole file (indentation changes because module does not use brackets `{...}` anymore).

muXxer commented on crates/iota-framework/packages/iota-framework/sources/test/test_utils.move:
```diff
@@ -19,6 +19,92 @@ module iota::test_utils {
+
+            while (j < vector::length(&unique_values)) {
+                if (unique_values[j] == value) {
+                    let count = counts2[j];
```
why is this line different here than the first loop for `t1` ?
There we use `let count = *vector::borrow(&counts1, j);`

piotrm50 replied:
Changed to be consistent.

muXxer commented on crates/iota-framework/packages/iota-framework/sources/test/test_utils.move:
```diff
@@ -19,6 +19,92 @@ module iota::test_utils {
         }
     }
 
+    public fun assert_same_elems<T: drop + copy>(t1: vector<T>, t2: vector<T>) {
```
can we add a description here? I would expect that it checks that all elements are the same at the same position, but from the code I understand that they don't need to be in the same order, is that intended?

piotrm50 replied:
Yes, to check if the elems are the same at the same position I could use a built-in `.equals()`. I will add a description.

muXxer commented on crates/iota-framework/packages/iota-system/sources/iota_system_state_inner.move:
```diff
@@ -20,12 +20,49 @@ module iota_system::iota_system_state_inner {
     const SYSTEM_STATE_VERSION_V1: u64 = 1;
 
+    /// A list of system config parameters.
+    public struct SystemParametersV2 has store {
```
Shouldn't we move the V2 after the definition of V1?
We also should add a description was was the reason to introduce V2 compared to V1. (new fields etc).

piotrm50 replied:
done.

muXxer commented on crates/iota-framework/packages/iota-system/sources/iota_system_state_inner.move:
```diff
@@ -296,14 +333,37 @@ module iota_system::iota_system_state_inner {
+                epoch_duration_ms,
+                min_validator_count,
+                max_validator_count,
+                committee_members_count: 20, // TODO: what should the initial value be?
```
should that actually be hardcoded? can't we pass that to the genesis?

piotrm50 replied:
Genesis uses`SystemParametersV1`, so when the network is already in operation there is no other source of those values here. In Sui's codebase a new parameter `min_validator_count` is also hardcoded in the upgrade function. 
The only question is what should this value be set to.

muXxer replied:
We should rename it to `max_committee_members_count`.

piotrm50 replied:
Renamed

muXxer commented on crates/iota-framework/packages/iota-system/tests/voting_power_tests.move:
```diff
@@ -13,9 +13,8 @@ module iota_system::voting_power_tests {
-        voting_power::set_voting_power(&mut validators);
-        test_utils::assert_eq(get_voting_power(&validators), voting_power);
+        let (committee_members, mut validators) = gtu::create_validators_with_stakes(stakes, ctx);
+        voting_power::set_voting_power_v2(&committee_members, &mut validators);        test_utils::assert_eq(get_voting_power(&validators), voting_power);
```
why is that in one line?

piotrm50 replied:
lack of formatter..

muXxer commented on crates/iota-types/src/iota_system_state/iota_system_state_summary.rs:
```diff
@@ -396,10 +407,27 @@ impl IotaSystemStateSummaryV1 {
+    }
+
     pub fn get_iota_committee_for_benchmarking(&self) -> CommitteeWithNetworkMetadata {
         let validators = self
```
Maybe rename to `committee` ?

piotrm50 replied:
done

muXxer commented on crates/iota-types/src/iota_system_state/iota_system_state_inner_v2.rs:
```diff
@@ -100,8 +162,7 @@ impl IotaSystemStateTrait for IotaSystemStateV2 {
     fn get_current_epoch_committee(&self) -> CommitteeWithNetworkMetadata {
         let validators = self
```
Maybe rename to `committee` ?

muXxer commented on crates/iota-types/src/iota_system_state/iota_system_state_inner_v2.rs:
```diff
@@ -26,6 +26,68 @@ use crate::{
 
+/// Rust version of the Move iota::iota_system::SystemParametersV2 type
+#[derive(Debug, Serialize, Deserialize, Clone, Eq, PartialEq)]
+pub struct SystemParametersV2 {
```
Can you add a description what changed compared to V1?

piotrm50 replied:
This struct will be removed after latest discussion.

alexsporn commented on crates/iota-framework/packages/iota-system/sources/iota_system_state_inner.move:
```diff
@@ -821,18 +881,19 @@ module iota_system::iota_system_state_inner {
 
         // Derive the reference gas price for the new epoch
-        self.reference_gas_price = self.validators.derive_reference_gas_price();
+        self.reference_gas_price = self.validators.derive_reference_gas_price_inner();
```
should we clean up all the gas price survey with this change?

piotrm50 replied:
Was there a reason this wasn't removed with PCR2? Maybe we should do it with protocol-version-5, but definitely on a separate PR.

piotrm50 replied:
remove this line

piotrm50 replied:
removed

muXxer commented on crates/iota-types/src/iota_system_state/iota_system_state_inner_v2.rs:
```diff
@@ -26,6 +26,68 @@ use crate::{
+}
+
+/// Rust version of the Move iota_system::validator_set::ValidatorSetV2 type
+#[derive(Debug, Serialize, Deserialize, Clone, Eq, PartialEq)]
```
Can you add a description was changed compared to V1?

piotrm50 replied:
sure, done

alexsporn commented on crates/iota-framework/packages/iota-system/sources/voting_power.move:
```diff
@@ -31,26 +31,46 @@ module iota_system::voting_power {
     /// Each validator's voting power is initialized using their stake. We then attempt to cap their voting power
     /// at `MAX_VOTING_POWER`. If `MAX_VOTING_POWER` is not a feasible cap, we pick the lowest possible cap.
-    public(package) fun set_voting_power(validators: &mut vector<ValidatorV1>) {
+    public(package) fun set_voting_power_v1(validators: &mut vector<ValidatorV1>) {
```
why do we need this v1?

piotrm50 replied:
This is used during IotaSystemStateV1 during genesis creation.

alexsporn commented on crates/iota-framework/packages/iota-system/sources/voting_power.move:
```diff
@@ -31,26 +31,46 @@ module iota_system::voting_power {
     /// descending order using voting power.
     /// Anything beyond the threshold is added to the remaining_power, which is also returned.
-    fun init_voting_power_info(
+    fun init_voting_power_info_v1(
```
same here, why do we need a v1?

piotrm50 replied:
Used in `set_voting_power_v1` which is then indirectly used during genesis creation.

alexsporn commented on crates/iota-framework/packages/iota-system/sources/voting_power.move:
```diff
@@ -120,7 +199,31 @@ module iota_system::voting_power {
 
     /// Update validators with the decided voting power.
-    fun update_voting_power(validators: &mut vector<ValidatorV1>, mut info_list: vector<VotingPowerInfoV1>) {
+    fun update_voting_power_v1(validators: &mut vector<ValidatorV1>, mut info_list: vector<VotingPowerInfoV1>) {
```
Same here, why v1?

piotrm50 replied:
Indirectly used in genesis creation.

alexsporn commented on crates/iota-framework/packages/iota-system/sources/voting_power.move:
```diff
@@ -132,9 +235,55 @@ module iota_system::voting_power {
 
     /// Check a few invariants that must hold after setting the voting power.
-    fun check_invariants(v: &vector<ValidatorV1>) {
+    fun check_invariants_v1(v: &vector<ValidatorV1>) {
```
Same here, why v1?

piotrm50 replied:
Indirectly used in genesis creation.

muXxer commented on crates/iota-framework/packages/iota-system/tests/validator_set_tests.move:
```diff
@@ -79,17 +96,128 @@ module iota_system::validator_set_tests {
+        advance_epoch_with_dummy_rewards(&mut validator_set, committee_size, scenario);
+
+
+        // Validator6 joins the committee and brings 900 worth of stake after its stake increased, replacing Validator2 who has the lowest stake.
```
But the stake from Validator2 isn't gone then?

piotrm50 replied:
Fixed the comment so it's easier to understand.

muXxer commented on crates/iota-framework/packages/iota-system/tests/validator_set_tests.move:
```diff
@@ -79,17 +96,128 @@ module iota_system::validator_set_tests {
 
+    #[test]
+    fun test_top_stakers_committee_selection() {
+        // Create 5 validators with different stakes.
```
I count 9 :D

muXxer commented on crates/iota-framework/packages/iota-system/tests/validator_set_tests.move:
```diff
@@ -79,17 +96,128 @@ module iota_system::validator_set_tests {
+
+        let committee_size = 5;
+
+        // Create a validator set with only the first validator in it.
```
The first 4 are in it.

muXxer commented on crates/iota-framework/packages/iota-system/tests/validator_set_tests.move:
```diff
@@ -79,17 +96,128 @@ module iota_system::validator_set_tests {
+        assert_same_elems(validator_set.committee_validator_addresses(),  vector[@0x1, @0x2, @0x3, @0x4]);
+        assert_eq(validator_set.total_stake_inner(), 20 * 100 * NANOS_PER_IOTA);
+
+        add_and_activate_validator(&mut validator_set, v5, scenario);
```
It's kinda self explanatory, but I would still add comments for the intentions in this test.
For example:

"add 5th validator and advance to new epoch"
"check if 5th validator is in the committee"

piotrm50 replied:
Added comments to the whole test.

muXxer commented on crates/iota-framework/packages/iota-system/tests/validator_set_tests.move:
```diff
@@ -79,17 +96,128 @@ module iota_system::validator_set_tests {
+
+        assert_same_elems(validator_set.active_validator_addresses(), vector[@0x1, @0x2, @0x3, @0x4, @0x5]);
+        assert_same_elems(validator_set.committee_validator_addresses(),  vector[@0x1, @0x2, @0x3, @0x4, @0x5]);
+        assert_eq(validator_set.total_stake_inner(), 40 * 100 * NANOS_PER_IOTA);
```
Where does the new stake come from? Can we add a comment why it's 20 more each time?

piotrm50 replied:
Added comments to the whole test so that it's clear.

muXxer commented on crates/iota-framework/packages/iota-system/tests/validator_set_tests.move:
```diff
@@ -79,17 +96,128 @@ module iota_system::validator_set_tests {
+        advance_epoch_with_dummy_rewards(&mut validator_set, committee_size, scenario);
+
+        assert_same_elems(validator_set.active_validator_addresses(), vector[@0x1, @0x2, @0x3, @0x4, @0x5, @0x6]);
+        assert_same_elems(validator_set.committee_validator_addresses(),  vector[@0x2, @0x3, @0x4, @0x5, @0x6]);
```
add a comment that first drops out because of max 5 validators and lowest stake.

piotrm50 replied:
Added comments to the whole test so that it's clear.

muXxer commented on crates/iota-framework/packages/iota-system/tests/validator_set_tests.move:
```diff
@@ -79,17 +96,128 @@ module iota_system::validator_set_tests {
+        advance_epoch_with_dummy_rewards(&mut validator_set, committee_size, scenario);
+
+        assert_same_elems(validator_set.active_validator_addresses(), vector[@0x1, @0x2, @0x3, @0x4, @0x5, @0x6, @0x7]);
+        assert_same_elems(validator_set.committee_validator_addresses(),  vector[@0x3, @0x4, @0x5, @0x6, @0x7]);
```
same here.

piotrm50 replied:
Added comments to the whole test so that it's clear.

muXxer commented on crates/iota-framework/packages/iota-system/tests/validator_set_tests.move:
```diff
@@ -79,17 +96,128 @@ module iota_system::validator_set_tests {
+        advance_epoch_with_dummy_rewards(&mut validator_set, committee_size, scenario);
+
+        assert_same_elems(validator_set.active_validator_addresses(), vector[@0x1, @0x2, @0x3, @0x4, @0x5, @0x6, @0x7, @0x8]);
+        assert_same_elems(validator_set.committee_validator_addresses(),  vector[@0x3, @0x4, @0x5, @0x6, @0x7]);
```
add a comment that v8 is not added to the committee because the stake is lower than the rest.

piotrm50 replied:
Done

muXxer commented on crates/iota-framework/packages/iota-system/tests/validator_set_tests.move:
```diff
@@ -79,17 +96,128 @@ module iota_system::validator_set_tests {
     #[test]
     fun test_reference_gas_price_derivation() {
+        // TODO: implement tests that check that gas price is correctly calculated based on committee and not all validators
         // Create 5 validators with different stakes and different gas prices.
```
it's 10 validators.

muXxer commented on crates/iota-framework/packages/iota-system/tests/validator_set_tests.move:
```diff
@@ -99,42 +227,103 @@ module iota_system::validator_set_tests {
+        );
+        advance_epoch_with_dummy_rewards(&mut validator_set, committee_size, scenario);
+
+        assert_eq(validator_set.derive_reference_gas_price_inner(), 42);
```
can't we add an explanation why the gas price is this value?
It's hard to understand, and I guess updating those tests will be a pain if we don't know why it is that value.
same for all the following `derive_reference_gas_price_inner` assertions.

piotrm50 replied:
Gas price survey will be removed entirely.

muXxer commented on crates/iota-framework/packages/iota-system/tests/validator_set_tests.move:
```diff
@@ -99,42 +227,103 @@ module iota_system::validator_set_tests {
+        assert_eq(validator_set.derive_reference_gas_price_inner(), 43);
+        
+        // Start adding more validators than the committee can hold and 
+        // make sure that only committee members take part in the gas price survey.
```
Can we add a comment before `add_and_activate_validator` if this validator takes part in the survey? Or where is this info written?

piotrm50 replied:
This logic will be removed entirely.

muXxer commented on crates/iota-framework/packages/iota-system/tests/validator_set_tests.move:
```diff
@@ -361,9 +558,10 @@ module iota_system::validator_set_tests {
         );
-        assert_eq(active_validator_addresses(&validator_set), vector[@0x3, @0x4]);
+        assert_same_elems(active_validator_addresses(&validator_set), vector[@0x3, @0x4]);
+        assert_same_elems(committee_validator_addresses(&validator_set), vector[ @0x3, @0x4]);
```
```suggestion
        assert_same_elems(committee_validator_addresses(&validator_set), vector[@0x3, @0x4]);
```

muXxer commented on crates/iota-framework/packages/iota-system/tests/validator_set_tests.move:
```diff
@@ -8,11 +8,11 @@ module iota_system::validator_set_tests {
     use iota::coin;
     use iota_system::staking_pool::StakedIota;
     use iota_system::validator::{Self, ValidatorV1, staking_pool_id};
-    use iota_system::validator_set::{Self, ValidatorSetV1, active_validator_addresses};
```
I don't see any tests that dynamically change the size of the committee?
Shouldn't that also be tested, in case the protocol parameter changes during epochs?

piotrm50 replied:
True, need to add that!

muXxer commented on crates/iota-framework/packages/iota-system/sources/voting_power.move:
```diff
@@ -31,26 +31,46 @@ module iota_system::voting_power {
     const EVotingPowerOverThreshold: u64 = 3;
     const EInvalidVotingPower: u64 = 4;
 
+    const ECommitteeMembersSetCorrupt: u64 = 102;
```
why `102` ?

muXxer replied:
Ah I see, because that was also used in another file. But should the error codes even be kept in sync? 🤷

piotrm50 replied:
It doesn't need to be kept in sync, but I wanted to avoid confusion of having 2 error codes with exactly the same meaning in a very close context.

muXxer commented on crates/iota-framework/packages/iota-system/sources/voting_power.move:
```diff
@@ -31,26 +31,46 @@ module iota_system::voting_power {
+        check_invariants_v1(validators);
+    }
+
+    /// Set the voting power of all validators.
```
Shouldn't the comment be altered (also inside the function)?
Isn't it more about the committee members?

piotrm50 replied:
Good point.

muXxer commented on crates/iota-framework/packages/iota-system/sources/voting_power.move:
```diff
@@ -88,6 +147,26 @@ module iota_system::voting_power {
+                ECommitteeMembersSetCorrupt,
+            );
+
+            let v = &validators[validator_index];
```
In the function `init_voting_power_info_v2` we call it `validator` instead of `v`, but in the following functions we call it `v`. Seems a bit inconsistent all over the file :trollface:

piotrm50 replied:
True, that's how we inherited already. Should I fix naming everywhere or should we do it in a separate PR after that to avoid cluttering this PR with renames?

muXxer replied:
Lets fix it later.

muXxer commented on crates/iota-framework/packages/iota-system/sources/voting_power.move:
```diff
@@ -132,9 +235,55 @@ module iota_system::voting_power {
         info_list.destroy_empty();
     }
+    /// Check a few invariants that must hold after setting the voting power.
+    fun check_invariants_v2(committee_members: &vector<u64>, v: &vector<ValidatorV1>) {
```
Move v2 after v1.

muXxer commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -58,6 +58,47 @@ module iota_system::validator_set {
         extra_fields: Bag,
     }
 
+    public struct ValidatorSetV2 has store {
```
Add a description what has changed compared to V1.

piotrm50 replied:
done

muXxer commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -115,14 +173,50 @@ module iota_system::validator_set {
 
-    public(package) fun new(init_active_validators: vector<ValidatorV1>, ctx: &mut TxContext): ValidatorSetV1 {
-        let total_stake = calculate_total_stakes(&init_active_validators);
+    public(package) fun import(
```
Isn't `import` a weird name? Can't we have something more descriptive?
`new_v2_from_whatever`

piotrm50 replied:
I'm actually wondering why this import is necessary since it's only called in `v1_to_v2` method. I'll try to put the code directly in the place where it's called.

muXxer commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -115,14 +173,50 @@ module iota_system::validator_set {
+            extra_fields,
+        };
+
+        validators
```
Is there no need to recalculate stuff if we "import" ?

In the functions `new_v1` and `new_v2` we set the voting power after creating the `ValidatorSetVx`. Isn't this needed here?

piotrm50 replied:
`v1_to_v2` should not select a new committee or change anything about it. It should just make it so that we use a new struct without changing anything about the committee. A new committee will be selected later, together with assigning weights etc.

`new_v1` and `new_v1` are creating a completely new instance, so the fields need to be initialized, but here we're reusing existing ValidatorSet that has all the fields initialized already.

muXxer commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -509,8 +707,37 @@ module iota_system::validator_set {
     }
 
+    // ==== upgradeable getter functions for ValidatorSetV2 ====
+    public(package) fun total_stake_inner(self: &ValidatorSetV2): u64 {
```
Maybe stupid question, but why do we call all of them `inner`?
What happens if we create a `V3` at some point?

piotrm50 replied:
The problem is that `public fun total_stake(self: &ValidatorSetV1)` is `public`, so we can't change the signature to pass `ValidatorSetV2`. 
The `inner` methods are `public(package)`, which you can modify in the future when we create `V3`. 

That's why you see that some methods (that are `public(package)`) have updated signatures, while some are not touched and turned into those `inner` funcs.

muXxer commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -525,24 +752,33 @@ module iota_system::validator_set {
         find_validator(&self.active_validators, validator_address).is_some()
     }
 
+    /// Returns true iff the address exists in committee validators.
```
```suggestion
    /// Returns true if the address exists in the committee validators.
```

piotrm50 replied:
`iff` is not wrong and present in other comments - `if and only if`

muXxer commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -525,24 +752,33 @@ module iota_system::validator_set {
     // ==== private helpers ====
 
     /// Checks whether `new_validator` is duplicate with any currently active validators.
-    /// It differs from `is_active_validator_by_iota_address` in that the former checks
```
Why was the comment removed? Isn't it still true?

piotrm50 replied:
true, added it back

muXxer commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -709,13 +945,24 @@ module iota_system::validator_set {
+        validator_address: address,
+    ): &ValidatorV1 {
+        let mut validator_index_opt = find_validator(validators, validator_address);
+        assert!(validator_index_opt.is_some() && committee_members.contains(validator_index_opt.borrow()), ENotACommitteeValidator);
```
Shouldn't we split the two checks, so we can also say that the validator address isn't even a validator in the first place, or would that make the logic on the calling site more complicated?

piotrm50 replied:
Good idea!

muXxer commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -747,26 +994,56 @@ module iota_system::validator_set {
+        validator_address: address,
+    ): &ValidatorV1 {
+        let mut validator_index_opt = find_validator(&self.active_validators, validator_address);
+        assert!(validator_index_opt.is_some() && self.committee_members.contains(validator_index_opt.borrow()), ENotACommitteeValidator);
```
Shouldn't we split the two checks, so we can also say that the validator address isn't even a validator in the first place, or would that make the logic on the calling site more complicated?

piotrm50 replied:
If assert fails then the program fails, because something went terribly wrong - the caller doesn't need to handle it.

Split the asserts.

muXxer commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -747,26 +994,56 @@ module iota_system::validator_set {
     /// Verify the capability is valid for a Validator.
-    /// If `active_validator_only` is true, only verify the Cap for an active validator.
+    /// If `committee_validator_only` is true, only verify the Cap for an committee validator.
     /// Otherwise, verify the Cap for au either active or pending validator.
```
"au"? Do they mean "an"?

muXxer commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -747,26 +994,56 @@ module iota_system::validator_set {
 
     /// Verify the capability is valid for a Validator.
-    /// If `active_validator_only` is true, only verify the Cap for an active validator.
+    /// If `committee_validator_only` is true, only verify the Cap for an committee validator.
```
```suggestion
    /// If `which_validator == COMMITTEE_VALIDATOR_ONLY` is true, only verify the Cap for an committee validator.
```

muXxer commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -1102,11 +1433,26 @@ module iota_system::validator_set {
     }
 
+
+    /// Sum up the total stake of a given list of validator addresses.
```
Please alter the comment, it's a copy paste from `sum_voting_power_by_addresses`

muXxer commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -1129,4 +1490,153 @@ module iota_system::validator_set {
+
+
+    // Selects top N stakers among all active validators to be part of the committee.
+    public(package) fun top_n_stakers(self: &mut ValidatorSetV2, n: u64) {
```
I would maybe call it `select_committee_members_top_n_stakers` or something?
Or at least `select_top_n_stakers`.

piotrm50 replied:
Good point.

semenov-vladyslav replied:
Or just make this function `top_n_stakers(self: &ValidatorSetV2, n: u64): vector<u64>` and make the final assignment at the call site, there's just two places where this function gets called.

muXxer commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -1129,4 +1490,153 @@ module iota_system::validator_set {
+
+    // Selects new set committee members from the set of all active validators.
+    // Currently selects all active validators to be part of the committee.
+    public(package) fun select_committee_members(self: &mut ValidatorSetV2) {
```
Is this even used somewhere? Don't we want to directly use `top_n_stakers`?

piotrm50 replied:
Right, removed it.

muXxer commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -1129,4 +1490,153 @@ module iota_system::validator_set {
+
+        // Sort indices based on the stake values 
+        let mut i = 1;
+        while (i < validators_num) {
```
Is insertion sort fine complexity wise in this case: O(n²)?
Or can we somehow implement Quicksort: O(n log n) on average case?

piotrm50 replied:
That's an improvement we can work on. The number of validators will not be huge, so it will not have a big impact.

semenov-vladyslav replied:
Why not use priority_queue?

piotrm50 replied:
Priority queue can only work with a single value and I've just implemented tie-breaking by comparing authority_key in case of equal stake. Unfortunately tuples are not comparable like in Rust.

muXxer commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -1129,4 +1490,153 @@ module iota_system::validator_set {
+        // Return the top N indices
+        let mut top_n_indices = vector[];
+        let mut i = 0;
+        while (i < n.min(validators_num)) {
```
Does it matter if the validators are sorted at the end?
If not, in case `validators_num <= n`, we could skip the sorting then.

piotrm50 replied:
Ah, that's smart. Will add that check.

muXxer commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -1129,4 +1490,153 @@ module iota_system::validator_set {
+        self.committee_members = top_n_indices;
+    }
+
+    // Emits events for committee validators that were added and left the committee.
```
```suggestion
    // Emits events for committee validators that were added or left the committee.
```

muXxer commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -1129,4 +1490,153 @@ module iota_system::validator_set {
+            if (!new_committee_addresses.contains(&validator_address)) {
+                let mut validator_index_opt = find_validator(&self.active_validators, validator_address);
+                
+                // If it's not part of active validators anymore, it means that the leave committee event has been emitted before.
```
Is that really the case? Can't it happen that the validator somehow gets removed from the `active_validators` set in another way? Maybe by deregistering as a validator or punishment by other validators? Idk.

piotrm50 replied:
Exactly, if the validator left the the committee by being removed from `active_validators`, then the event is emitted `process_validator_departure` function.

semenov-vladyslav commented on crates/iota-framework/packages/iota-framework/sources/test/test_utils.move:
```diff
@@ -19,6 +19,92 @@ module iota::test_utils {
+
+
+        // Vectors to store unique elements and their counts
+        let mut unique_values = vector<T>[];
```
No need to keep copies of T, just keep indices of (the first occurrence of) unique elements in t1.

piotrm50 replied:
good point!

semenov-vladyslav commented on crates/iota-framework/packages/iota-system/sources/voting_power.move:
```diff
@@ -31,26 +31,46 @@ module iota_system::voting_power {
+    public(package) fun set_voting_power_v2(committee_members: &vector<u64>, active_validators: &mut vector<ValidatorV1>) {
+        // If threshold_pct is too small, it's possible that even when all validators reach the threshold we still don't
+        // have 100%. So we bound the threshold_pct to be always enough to find a solution.
+        let threshold = TOTAL_VOTING_POWER.min(
```
The outer `TOTAL_VOTING_POWER.min` doesn't make sense: the value of the inner expression is always not greater than `TOTAL_VOTING_POWER` (unless `MAX_VOTING_POWER > TOTAL_VOTING_POWER` for some reason).

piotrm50 replied:
That's right, I copied this code from the `_v1` function. I'll leave it here - it's not a big improvement and it was added there for some reason in the first place.

piotrm50 commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -479,7 +640,44 @@ module iota_system::validator_set {
+    /// Derive the reference gas price based on the gas price quote submitted by each committee validator.
+    /// The returned gas price should be greater than or equal to 2/3 of the validators submitted
+    /// gas price, weighted by voting power in the current epoch.
+    public(package) fun derive_reference_gas_price_inner(self: &ValidatorSetV2): u64 {
```
Remove this function entirely.

piotrm50 committed:
Apply review suggestions

semenov-vladyslav commented on crates/iota-framework/packages/iota-system/sources/voting_power.move:
```diff
@@ -132,9 +235,55 @@ module iota_system::voting_power {
+                if (stake_a < stake_b) {
+                    assert!(power_a <= power_b, ERelativePowerMismatch);
+                };
+                b = b + 1;
```
And if the stakes are equal the voting powers should differ by not more than 1, otherwise the distribution is not fair.

piotrm50 replied:
Interesting, perhaps we should consider adding this check as well, but this is definitely out of scope of this PR - logic of this function was simply copied and adjusted for committee.

semenov-vladyslav commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -479,7 +640,44 @@ module iota_system::validator_set {
+                ECommitteeMembersSetCorrupt,
+            );
+
+            let v = &vs[validator_index];
```
Is it possible to hide this pattern (getting validator by index including the assert above) somehow, eg. in a macro fun? It is repeated a lot

piotrm50 replied:
Good point! Replaced it with a function.

Dkwcs commented on crates/iota-types/src/iota_system_state/iota_system_state_inner_v2.rs:
```diff
@@ -26,6 +26,68 @@ use crate::{
+#[derive(Debug, Serialize, Deserialize, Clone, Eq, PartialEq)]
+pub struct ValidatorSetV2 {
+    pub total_stake: u64,
+    pub active_validators: Vec<ValidatorV1>,
```
Could we have a comment here, that `active_validators` are validators with a chance to be selected to take part in consensus(not validators that actively taking part in consensus)?

piotrm50 replied:
Done.

Dkwcs commented on crates/iota-types/src/iota_system_state/iota_system_state_inner_v2.rs:
```diff
@@ -26,6 +26,68 @@ use crate::{
+pub struct ValidatorSetV2 {
+    pub total_stake: u64,
+    pub active_validators: Vec<ValidatorV1>,
+    pub committee_members: Vec<u64>,
```
And a comment here as well, that  `committee_members` are active validators indices that take active part in consensus(something like that)

piotrm50 replied:
Done.

piotrm50 committed:
Apply review suggestions.

piotrm50 committed:
`max_committee_members_count` is passed via AdvanceEpoch transaction instead of being stored in SystemParameters

gokhan-simsek-iota commented on crates/iota-framework/packages/iota-system/tests/governance_test_utils.move:
```diff
@@ -125,9 +129,11 @@ module iota_system::governance_test_utils {
         let ctx = scenario.ctx();
-
+        
+        // Use default value for max_active_validators.
```
for max_committee_members?

piotrm50 replied:
Improved comments

gokhan-simsek-iota commented on crates/iota-framework/packages/iota-system/tests/governance_test_utils.move:
```diff
@@ -162,8 +168,12 @@ module iota_system::governance_test_utils {
 
         let validator_subsidy = computation_charge;
+
+        // Use default value for max_active_validators.
```
for max_committee_members?

piotrm50 replied:
Improved the comment.

gokhan-simsek-iota commented on crates/iota-types/src/iota_system_state/iota_system_state_summary.rs:
```diff
@@ -484,6 +506,8 @@ impl From<IotaSystemStateSummaryV1> for IotaSystemStateSummaryV2 {
             validator_low_stake_grace_period,
             total_stake,
+            // All active validators are members of the committee.
+            committee_members: (0..active_validators.len() as u64).collect(),
```
So, the `committee_members` is the subset of active validators, why all of them are committee members here?

muXxer replied:
Yeah this is fine here in that case, because in the V1, all the validators were the "committee" basically.

kodemartin commented on crates/iota-rest-api/src/system.rs:
```diff
@@ -453,13 +464,15 @@ impl From<iota_types::iota_system_state::iota_system_state_summary::IotaSystemSt
             safe_mode_non_refundable_storage_fee,
             epoch_start_timestamp_ms,
             epoch_duration_ms,
+            max_committee_members_count,
```
This is a field in `AdvanceEpochParams` not in `IotaSystemStateSummaryV2`. Is this  a leftover from a previous iteration?

kodemartin replied:
Ok, I see this has been changed with https://github.com/iotaledger/iota/pull/4793/commits/23e1d7acc11951c86f0e854abbf02d01de42642e.

kodemartin replied:
Do you plan to refactor `iota-rest-api` and `iota-open-rpc` accordingly?

kodemartin replied:
Refactored with
c0c895bbf9
2e027f8060

kodemartin committed:
refactor(rest-api): remove max_committee_members_count from system summary

kodemartin committed:
chore(open-prc): update spec

kodemartin commented on iota-execution/latest/iota-adapter/src/execution_engine.rs:
```diff
@@ -881,6 +881,7 @@ mod checked {
             CallArg::Pure(bcs::to_bytes(&params.non_refundable_storage_fee).unwrap()), /* non_refundable_storage_fee: u64 */
             CallArg::Pure(bcs::to_bytes(&params.reward_slashing_rate).unwrap()), /* reward_slashing_rate: u64 */
             CallArg::Pure(bcs::to_bytes(&params.epoch_start_timestamp_ms).unwrap()), /* epoch_start_timestamp_ms: u64 */
+            CallArg::Pure(bcs::to_bytes(&params.max_committee_members_count).unwrap()), /* max_committee_members_count: u64 */
```
This looks ok, and aligns with the call args expected for `iota_system::iota_system::advance_epoch` Move function that is called through the PT constructed with `construct_advance_epoch_pt_impl`.

But I would expect the same arguments to be passed to `construct_advance_epoch_pt_impl` within `construct_advance_epoch_pt_v1`.

Why do we have two `construct_advance_epoch_pt_{v1,v2}`, if there is only as single `advance_epoch` Move function?

miker83z replied:
`construct_advance_epoch_pt_v1` is used by `advance_epoch_v1` which is used only for the `ChangeEpoch` transaction type. This transaction type is used by `iota-core` to construct a new change epoch transaction only when it finds itself in protocol versions 1 to 4. These have an iota-framework snapshot with the older version of the `advance_epoch` Move function. 

From protocol version 5, the core uses the `ChangeEpoch2` transaction type which makes use of the `advance_epoch` Move function you refer to (the new one).

kodemartin replied:
Ah, right, right!

kodemartin committed:
chore(framework): update compiled packages and snapshot manifest (#5966)

* chore(framework): update compiled packages

* chore(framework-snapshot): update manifest.json

miker83z commented on crates/iota-framework/packages/iota-system/sources/iota_system.move:
Could you revert the formatting of this file?

piotrm50 replied:
Damn, how did I miss this.

piotrm50 replied:
Sure! I don't know how I committed this. Diff must've skipped white spaces.

miker83z commented on crates/iota-framework/packages/iota-system/sources/iota_system_state_inner.move:
```diff
@@ -178,11 +178,11 @@ module iota_system::iota_system_state_inner {
     }
 
-    
+
```
```suggestion
```

miker83z commented on crates/iota-framework/packages/iota-system/sources/iota_system_state_inner.move:
```diff
@@ -296,14 +296,36 @@ module iota_system::iota_system_state_inner {
         // all computation charges are burned in protocol v1.
         let safe_mode_computation_charges_burned = safe_mode_computation_rewards.value();
+
+        let SystemParametersV1 {
+            epoch_duration_ms,
+            min_validator_count,
+            max_validator_count,
+            min_validator_joining_stake,
+            validator_low_stake_threshold,
+            validator_very_low_stake_threshold,
+            validator_low_stake_grace_period,
+            extra_fields: param_extra_fields,
+        } = parameters;
```
Why do you need to unwrap these here?

piotrm50 replied:
Good catch, leftover from some intermediate state.

kodemartin reviewed (approved):
looks good so far! I will follow up with separate PRs for `iota-{indexer,graphql-rpc,graphql-e2e-tests}` targeting the highest between [consensus/feat/committee-selection-feature](https://github.com/iotaledger/iota/tree/consensus/feat/committee-selection-feature) and [feat/committee-selection-rebased](https://github.com/iotaledger/iota/tree/feat/committee-selection-rebased)

Left a comment on something that doesn't look right on `iota-execution`, but is unrelated to this PR, and probably benign.

miker83z commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -776,23 +1033,27 @@ module iota_system::validator_set {
     fun process_pending_removals(
-        self: &mut ValidatorSetV1,
+        self: &mut ValidatorSetV2,
+        committee_addresses: vector<address>,
```
no need to use this attribute, you can get the vector from `self`

piotrm50 replied:
Well, `active_validators` already contains added validators, so `committee_members` indices should be still correct, but I think it's better to be explicit with committee_addresses once we have dirty `active_validators`.

miker83z commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -380,24 +513,30 @@ module iota_system::validator_set {
         very_low_stake_threshold: u64,
         low_stake_grace_period: u64,
         validator_report_records: &mut VecMap<address, VecSet<address>>,
+        committee_addresses: vector<address>,
```
no need to use this attribute, you can get the vector from `self`

piotrm50 replied:
It does not work in this case, because the `active_validators` vector is already modified and the committee_members values might point to incorrect Validators. That's why I get the vector with committee addresses before modifying the `active_validators` and pass it here.

kodemartin committed:
chore(framework-snapshot): update manifest.json

We update the reference commit in the manifest because #5966 squashed the commit
referenced previously, and it is no longer part of the history.

gokhan-simsek-iota commented on crates/iota-types/src/iota_system_state/epoch_start_iota_system_state.rs:
```diff
@@ -61,7 +61,7 @@ impl EpochStartSystemState {
         epoch_start_timestamp_ms: u64,
         epoch_duration_ms: u64,
-        active_validators: Vec<EpochStartValidatorInfoV1>,
+        committee_validators: Vec<EpochStartValidatorInfoV1>,
```
In the most of the places, `active_validators` and `committee_validators` exist together, but here no `active_validators`, is this OK ?

piotrm50 replied:
Yeah, no method in the trait requires it, so I it should be ok, unless some other team needs it.

piotrm50 committed:
Implement additional committee selection tests.

miker83z commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -380,24 +513,30 @@ module iota_system::validator_set {
+        
+        self.select_committee_members_top_n_stakers(committee_size);
+
+        self.emit_committee_change_events(prev_committee_validator_addresses, ctx);
```
I argue that the commitee vector update should be tied to the emission of join/leave events. The current logic split into two separate functions could lead to more complexity in the future, when some updates will be needed. I suggest these changes:
- return `top_n_indices` as result of `select_committee_members_top_n_stakers` instead of modifying `self`.
- rename `emit_committee_change_events` into `process_new_committee` (or something similar)
- remove `prev_committee_validator_addresses` as parameter
- invoke `let new_committee = select_committee_members_top_n_stakers()` inside `process_new_committee`
- invoke `let prev_committee = self.committee_validator_addresses()` inside `process_new_committee`
- work with these to emit join/leave events
- at the end of the function set `self.committee_members = new_committee`

piotrm50 replied:
I can apply some changes regarding the event, however `prev_committee_validator_addresses` must stay, because when adding and removing new `active_validators` the `committee_members` vector will point to incorrect validators, and this happens way before new committee is selected.

miker83z commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -1129,4 +1451,132 @@ module iota_system::validator_set {
         res
     }
+
+    public(package) fun committee_validator_addresses(self: &ValidatorSetV2): vector<address> {
```
Passing a `committee_members` vector<u64> as parameter instead of using `&self.committee_members` would help with the suggestion made above.

piotrm50 replied:
Still does not help, because at this point `active_validators` vector is already modified, so the `committee_members` is pointing to incorrect validators.

miker83z commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -747,26 +941,89 @@ module iota_system::validator_set {
+    fun get_validator_by_committee_index(
+        validators: &vector<ValidatorV1>,
+        committee_members: &vector<u64>,
+        committee_member_index: u64,
```
No need to pass a vector for a function invoke here. Instead of `committee_members: &vector<u64>, committee_member_index: u64,` should be only `committee_index: u64`, where `committee_index = committee_members[committee_member_index]` was computed in the caller.

miker83z commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -747,26 +941,89 @@ module iota_system::validator_set {
+
+    fun get_validator_by_committee_index_mut(
+        validators: &mut vector<ValidatorV1>,
+        committee_members: &vector<u64>,
```
same here

miker83z commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -1129,4 +1451,132 @@ module iota_system::validator_set {
+            validator_indices.push_back(i);
+
+            i = i + 1;
+        };
```
```suggestion
        let mut validator_indices = vector::tabulate!(validators_num, |i| i);
```

miker83z commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -1129,4 +1451,132 @@ module iota_system::validator_set {
+        };
+
+        // Return the top N indices
+        let mut top_n_indices = vector[];
+        let mut i = 0;
+        while (i < n.min(validators_num)) {
+            top_n_indices.push_back(validator_indices[i]);
+            i = i + 1;
+        };
```
```suggestion
        let top_n_indices = vector::tabulate!(n.min(validators_num), |i| validator_indices[i]);
```

piotrm50 replied:
Ah, I knew something like this must exist!

piotrm50 committed:
Fix rewards distribution test

miker83z commented:
Regarding the versioning in `voting_power.move`, I think this can be avoided. Here's a proposal: #5984

miker83z committed:
refactor(iota-system): Remove versioning from voting power (#5984)

* refactor(iota-system): remove versioning from voting power

* refactor(iota-system): use of get_validator_by_committee_index and reduce error handling

kodemartin committed:
feat: expose committee members on indexer dependent services (#5975)

* feat!(json-rpc-types): add committee_members to EpochInfo

* chore(open-rpc): update spec with new EpochInfo

* refactor(indexer): require deserialization of stored system state

* feat(indexer): update EpochInfo with committee_members

* feat(graphql-rpc): expose committeeMembers as part of validator set

* chore(graphql-rpc): update schema and snapshot

* refactor(graphql-e2e-tests): apply fix and update baselines

* fixup! refactor(graphql-e2e-tests): apply fix and update baselines

muXxer committed:
fix: update protocol config snapshots

muXxer committed:
fix: update published_api + iota-system

muXxer committed:
fix: update iota-framework-snapshot

muXxer committed:
fix: update test snapshot

piotrm50 committed:
Undo accidental formatting

piotrm50 committed:
Apply review suggestions

Thoralf-M commented on crates/iota-framework/packages/iota-system/sources/validator.move:
```diff
@@ -608,6 +611,32 @@ module iota_system::validator {
+            i = i + 1;
+        };
+        
+        false // Should never end up here
```
Is there anything that makes sure of that for the authority_pubkey? I only see a check for the iota_address
https://github.com/iotaledger/iota/blob/474ca3a70dc536502f65d96ebaf85d5893158a69/crates/iota-framework/packages/iota-system/sources/validator_set.move#L164-L168
So maybe the ordering here should also use the `iota_address` instead?

piotrm50 replied:
Here we check that all the fields are unique. Validator will fail to be added it any of the fields are equal.

https://github.com/iotaledger/iota/blob/62274b80f683ffbe3ad0dce6b3045d1bd94fdaf0/crates/iota-framework/packages/iota-system/sources/validator.move#L562

Thoralf-M replied:
`MUSTFIX: We need to check this when updating metadata as well.` :sweat_smile:

Thoralf-M commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -1102,11 +1368,26 @@ module iota_system::validator_set {
         &self.active_validators
     }
-
+    
```
Weird empty spaces

piotrm50 replied:
removed

Thoralf-M commented on crates/iota-framework/packages/iota-system/tests/validator_set_tests.move:
```diff
@@ -79,62 +96,307 @@ module iota_system::validator_set_tests {
-        advance_epoch_with_dummy_rewards(&mut validator_set, scenario);
 
-        assert_eq(validator_set.derive_reference_gas_price(), 42);
+    // TODO: implement tests that check max_committee_members_count changes
```
What about this TODO?

piotrm50 replied:
Implemented the test, forgot to remove the todo.

piotrm50 committed:
Apply review suggestions again

piotrm50 committed:
Update the snapshot

bingyanglin committed:
Use committee_members for validators that take active part in consensus (#6007)

muXxer commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -58,6 +58,52 @@ module iota_system::validator_set {
+        total_stake: u64,
+
+        /// The current list of active validators.
+        active_validators: vector<ValidatorV1>,
```
@miker83z and I discussed that it might make sense to directly switch all occurences of `ValidatorV1` inside the new `ValidatorSetV2` to use the wrapper type `Validator`.

miker83z replied:
I tried an implementation and it won't work because at some point the tx context is needed. This can be passed in most invocations of the v1_to_v2, however at some point it clashes with some public methods of `IotaSystemState` (like [here](https://github.com/iotaledger/iota/blob/7818c1b3a68089f5e2a3ce3c8d50c8f015a0c38f/crates/iota-framework/packages/iota-system/sources/iota_system.move#L498)).

Maybe there's a solution, but it implies to deprecate the wrapper `Validator` as it is. Instead of using `Versioned`, the wrapper should be represented in a similar way to `IotaSystemState` and have methods like `load_system_state` and `load_system_state` that make use of dynamic fields internally. With dynamic fields you don't need a ctx.

piotrm50 replied:
Since this is not directly related to this PR I would suggest we work on this in a separate PR as not to block this one.

muXxer committed:
fix: update graphql-e2e-tests (#6009)

alexsporn commented on crates/iota-protocol-config/src/lib.rs:
```diff
@@ -1689,6 +1696,7 @@ impl ProtocolConfig {
                 5 => {
                     cfg.feature_flags.protocol_defined_base_fee = true;
                     cfg.base_gas_price = Some(1000);
+                    cfg.max_committee_members_count = Some(20);
```
The default should be 50

lzpap commented on iota-execution/latest/iota-adapter/src/execution_engine.rs:
```diff
@@ -1013,6 +1017,7 @@ mod checked {
             non_refundable_storage_fee: change_epoch_v2.non_refundable_storage_fee,
             reward_slashing_rate: protocol_config.reward_slashing_rate(),
             epoch_start_timestamp_ms: change_epoch_v2.epoch_start_timestamp_ms,
+            max_committee_members_count: protocol_config.max_committee_members_count(),
```
Advance epoch v2 was introduced with the base gas fee changes. Now we added a new parameter, can someone explain to me why we don't need v3? If you deploy this node sw on the network but protocol v5 has not commenced yet, won't the nodes call `advance_epoch` with one more argument that they should?

miker83z replied:
We still have no live networks with any epoch using the `EpochV2` so we are still in the safe zone when it comes to modifying the type.

lzpap replied:
fair enough

piotrm50 committed:
Fix protocol upgrade tests

piotrm50 committed:
Update default max_committee_members_count to 50

piotrm50 committed:
Update bytecode snapshot

piotrm50 committed:
Update git revision so spell checker does not complain

piotrm50 committed:
Fix expected files and expected snapshot

miker83z commented on crates/iota-framework/packages/iota-system/sources/validator.move:
```diff
@@ -608,6 +611,31 @@ module iota_system::validator {
+        let other_pubkey = other.authority_pubkey_bytes();
+
+        // Compare the two pubkeys lexicographically, assuming equal lengths
+        let len = self_pubkey.length();
+
+        let mut i = 0;
+        while (i < len) {
+            let self_byte = self_pubkey[i];
+            let other_byte = other_pubkey[i];
+
+            if (self_byte != other_byte) {
+                return self_byte < other_byte
+            };
+            i = i + 1;
+        };
+        
+        false // Should never end up here
```
```suggestion
        'smaller_than: {
            self_pubkey.zip_do_ref!(other_pubkey, |a, b| if (a != b) return 'smaller_than *a < *b);
            false // Should never end up here
        }
```

piotrm50 replied:
Applied

miker83z commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -479,7 +622,7 @@ module iota_system::validator_set {
     }
 
-    // ==== getter functions ====
+    // ==== getter functions for ValidatorSetV1 ====
```
Could you check if using the deprecated annotation in theve V1 public fun breaks something? This is the notation: 
```
#[deprecated]
```

piotrm50 replied:
added the deprecated annotation

miker83z commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -1102,6 +1365,19 @@ module iota_system::validator_set {
+        let mut i = 0;
+        let length = addresses.length();
+        while (i < length) {
+            let validator = get_committee_validator_ref(vs, committee_members, addresses[i]);
```
Given the usage of `sum_committee_voting_power_by_addresses`, this should invoke `get_committee_validator_ref_inner` and then the private method `get_committee_validator_ref` can be removed. 
Then, also we won't need the `committee_members` parameter for `sum_committee_voting_power_by_addresses`.

piotrm50 replied:
Refactored as suggested

miker83z commented on crates/iota-framework/packages/iota-system/sources/validator_set.move:
```diff
@@ -1117,7 +1393,22 @@ module iota_system::validator_set {
     }
 
-    public(package) fun active_validator_addresses(self: &ValidatorSetV1): vector<address> {
+    /// Return the active validators in `self`
```
Also, maybe we could also deprecate the functions above for V1

piotrm50 replied:
added the deprecated annotation

piotrm50 committed:
Apply review suggestions

piotrm50 committed:
Regenerate iota-system binaries

piotrm50 committed:
Update manifest and expected values for tests

miker83z reviewed (approved):
LGTM!

piotrm50 committed:
Merge remote-tracking branch 'origin/develop' into feat/committee-selection-rebased

piotrm50 committed:
Rename validator_address_by_pool_id to validator_address_by_pool_id_inner

piotrm50 committed:
Recompile iota-system

piotrm50 committed:
Update framework snapshot and expected values

piotrm50 committed:
rustfmt

begonaalvarezd committed:
chore(ts-sdk): update ts sdk to support committee selection protocol update (#6011)

* chore(ts-sdk): re-generate ts sdk client types

* chore(ts-sdk): update graphql auto-generated files

* chore(ts-sdk): update graphql auto-generated queries

* chore(ts-śdk): add and update committee info

* chore(ts-sdk): add changeset

* chore(ts-sdk): remove unnecesary fields from graphql queries

* chore(ts-sdk): update graphql paginated checkpoints

* chore(ts-sdk): Update checkpoint & getCommitteeInfo queries

* chore(ts-sdk): Update checkpoint

* chore(ts-sdk): Refactor rename checkpoint.ts

---------

Co-authored-by: Mario Sarcevic <mario.sarcevic@iota.org>

piotrm50 committed:
Fix validator tests
